src/VcfReader.cpp

```diff
@@ -0,0 +1,378 @@
+
+
+bool get_info_field(const string &info, const string &key, string &out) {
+    const uint8_t KEY_LENGTH = key.length()+1;
+    size_t i, p, q;
+
+    p=info.find(key+"=",0);
+    if (p==string::npos) return false;
+    if (key==END_STR && p>=2 && info.at(p-2)=='C' && info.at(p-1)=='I') {
+        p=info.find(key+"=",p+KEY_LENGTH);
+        if (p==string::npos) return false;
+    }
+    q=info.find(INFO_SEPARATOR,p+KEY_LENGTH);
+    if (q==string::npos) q=info.length();
+    out.clear();
+    for (i=p+KEY_LENGTH; i<q; i++) out+=info.at(i);
+    return true;
+}
+
+
+void ncalls_in_sample(const string& buffer, bool is_diploid, pair<uint8_t, uint8_t>& out) {
+    char c;
+    const size_t LENGTH = buffer.length();
+    size_t i, j;
+
+    out.first=0; out.second=0;
+    if (!is_diploid) {
+        if (buffer.starts_with('0')) out.first=1;
+        else if (!buffer.starts_with(VCF_MISSING_CHAR)) out.second=1;
+    }
+    else {
+        j=0;
+        for (i=0; i<LENGTH; i++) {
+            c=buffer.at(i);
+            if (c==PHASED_CHAR || c==UNPHASED_CHAR) {
+                if (buffer.starts_with('0')) out.first=1;
+                else if (!buffer.starts_with(VCF_MISSING_CHAR)) out.second=1;
+                j=i+1;
+            }
+            else if (c==GT_SEPARATOR) {
+                c=buffer.at(j);
+                if (c=='0') out.first++;
+                else if (c!=VCF_MISSING_CHAR) out.second++;
+                return;
+            }
+        }
+        c=buffer.at(j);
+        if (c=='0') out.first++;
+        else if (c!=VCF_MISSING_CHAR) out.second++;
+    }
+}
+
+
+bool is_phased(const string& buffer) {
+    const size_t LENGTH = buffer.length();
+
+    for (size_t i=0; i<LENGTH; i++) {
+        switch (buffer.at(i)) {
+            case PHASED_CHAR: return true;
+            case UNPHASED_CHAR: return false;
+            case GT_SEPARATOR: break;
+        }
+    }
+    return true;
+}
+
+
+void info2map(const string& info, unordered_map<string,string>& map, string& tmp_buffer_1, string& tmp_buffer_2) {
+    bool found;
+    char c;
+    uint16_t i;
+    const uint16_t LENGTH = info.length();
+
+    map.clear(); found=false;
+    tmp_buffer_1.clear(); tmp_buffer_2.clear();
+    for (i=0; i<LENGTH; i++) {
+        c=info.at(i);
+        if (c==INFO_ASSIGNMENT) found=true;
+        else if (c==INFO_SEPARATOR) {
+            map.emplace(tmp_buffer_1,tmp_buffer_2);
+            tmp_buffer_1.clear(); tmp_buffer_2.clear();
+            found=false;
+        }
+        else if (!found) tmp_buffer_1+=c;
+        else tmp_buffer_2+=c;
+    }
+    if (found) map.emplace(tmp_buffer_1,tmp_buffer_2);
+}
+
+
+void for_record_in_vcf(const string& path, const function<void(VcfRecord& record)>& f, uint32_t progress_n_lines, const string& chromosome, bool is_diploid, bool high_qual_only, float min_qual, bool pass_only, uint32_t min_sv_length, bool skip_samples, uint32_t n_samples_in_vcf, float min_allele_frequency, float min_nonmissing_frequency) {
```

Comment:
We probably shouldn't skip the `chrom` field, it's not going to be a bottleneck for the pipeline